### PR TITLE
bpo-42047: DragonFlyBSD thread native id support.

### PR DESCRIPTION
**Pull request by devnexen**
https://bugs.python.org/issue42047

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2020-10-15-18-37-12.bpo-42047.XDdoSF.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Adding thread native id to DragonflyBSD.
```
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Library/2020-10-15-18-37-12.bpo-42047.XDdoSF.rst]:
```suggestion
Added thread native id support for DragonflyBSD.
```

**Review by Fidget-Spinner — COMMENTED** [review]
LGTM, just one minor nitpick on news

**Review by Fidget-Spinner — APPROVED** [review]

**Comment by devnexen** [comment]
ping :)

File: Include/pythread.h [hunk 2]
```diff
@@ -20,7 +20,7 @@ PyAPI_FUNC(unsigned long) PyThread_start_new_thread(void (*)(void *), void *);
 PyAPI_FUNC(void) _Py_NO_RETURN PyThread_exit_thread(void);
 PyAPI_FUNC(unsigned long) PyThread_get_thread_ident(void);
 
-#if defined(__APPLE__) || defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) || defined(_WIN32) || defined(_AIX)
+#if defined(__APPLE__) || defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) || defined(__DragonFly__) || defined(_WIN32) || defined(_AIX)
```
Comment (erlend-aasland) [inline Include/pythread.h]: Ugh, we should break this up. @vstinner, are you ok with applying the following suggestion to this PR?

```suggestion
#if defined(__APPLE__)          \
    || defined(__linux__)       \
    || defined(__FreeBSD__)     \
    || defined(__OpenBSD__)     \
    || defined(__NetBSD__)      \
    || defined(__DragonFly__)   \
    || defined(_WIN32)          \
    || defined(_AIX)
#define PY_HAVE_THREAD_NATIVE_ID 
```

Comment (vstinner) [inline Include/pythread.h]: Wrapping on 80 columns makes the code more readable, but I don't think that it's worth it to put one OS per line. I suggest:

```suggestion
#if (defined(__APPLE__) || defined(__linux__) || defined(_WIN32) \
     || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) \
     || defined(__DragonFly__) || defined(_AIX))
```

I like to put surrounding parenthesis: ``(...)``.

Comment (erlend-aasland) [reply Include/pythread.h]: Ok; this is an improvement over the current formatting.

Comment (erlend-aasland) [reply Include/pythread.h]: Broken up by 5a54ad242fee3a3833c8478e51c2232328f6230a.

**Comment by erlend-aasland** [comment]
Please also update the `availability` annotation of `get_native_id` in the docs:

- `Doc/library/_thread.rst`
- `Doc/library/threading.rst`

See gh-13624 for inspiration :)

File: Misc/NEWS.d/next/Library/2020-10-15-18-37-12.bpo-42047.XDdoSF.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Added thread native id support for DragonflyBSD.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2020-10-15-18-37-12.bpo-42047.XDdoSF.rst]:
```suggestion
Add :func:`threading.get_native_id` support for DragonFly BSD.
```

File: Misc/NEWS.d/next/Library/2020-10-15-18-37-12.bpo-42047.XDdoSF.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Add :func:`threading.get_native_id` support for DragonFly BSD.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2020-10-15-18-37-12.bpo-42047.XDdoSF.rst]: Feel free to also credit yourself in the NEWS entry.

Comment (devnexen) [reply Misc/NEWS.d/next/Library/2020-10-15-18-37-12.bpo-42047.XDdoSF.rst]: Jesus I almost forgot about this one, thanks for the changes.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]
Please resolve https://github.com/python/cpython/pull/22714#issuecomment-1129885559

File: Doc/library/_thread.rst [hunk 5]
```diff
@@ -118,7 +118,7 @@ This module defines the following constants and functions:
    Its value may be used to uniquely identify this particular thread system-wide
    (until the thread terminates, after which the value may be recycled by the OS).
 
-   .. availability:: Windows, FreeBSD, Linux, macOS, OpenBSD, NetBSD, AIX.
+   .. availability:: Windows, FreeBSD, Linux, macOS, OpenBSD, NetBSD, AIX, DragonFlyBSD.
```
Comment (erlend-aasland) [inline Doc/library/_thread.rst:121]: Nit: is it DragonFly BSD or DragonFlyBSD? From their website, I suspect the former. OTOH, there is no consistent usage in the docs.

Comment (devnexen) [reply Doc/library/_thread.rst:121]: it s consistent with other BSD doc wise and does not stick out :-)

Comment (erlend-aasland) [reply Doc/library/_thread.rst:121]: I'm fine with that. If someone wants to normalise DragonFly BSD naming across the docs, feel free to take it on 😄

**Review by erlend-aasland — APPROVED** [review]

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, @devnexen!

**Pull request merged**